### PR TITLE
Pipe Spawners

### PR DESCRIPTION
**Pull request by TheTyrant25**
[OBJECTS] [INTERNAL] [MAPPING] [STATION SYSTEMS] [FEAT] [QOL]
## About the PR 
Hey, remember cable spawners? That, but it's pipes.
Disposal and Mail and Morgue and Cargo pipes, to be specific. Not atmospheric ones. We don't talk about those.

**BIG WARNING:** Pipes are inherently more complex and require manual placing of junctions. This is really just for drawing long stretches and 90 degree turns. DO NOT TRY TO MAKE 3 (or heaven forbid, 4) WAY JUNCTIONS WITH THESE. Place the three way junctions (like `/obj/disposalpipespawner/junction/left/east`) manually, then draw pipe spawners to and from them.

The pipe spawners will connect to pipes of their own type only, plus any junctions or smart switches or type sensing outlets or whatever.
- Introduces `/obj/disposalpipespawner`, with subtypes for all the inbuilt variants like 'cargo' and 'mail'.
- They automatically connect to each other, junctions, and regular pipes at runtime. They will replace themselves with either `/obj/disposalpipe/segment` or `/obj/disposalpipe/trunk` types, depending on their surroundings.
- Placing multiple pipe spawners of the same kind makes them delete themselves. Different kinds are able to cross without error.
- Attempting 3 or 4 way junctions with just pipe spawners causes a crash. So does solo pipe spawners.

_Before:_
Screenshot 2022-11-21 160454
_After:_
image
Pro tip: use the 'regular' subtypes for disposal pipes. The organisation of it was wacky before, but now as long as you use the `/obj/disposalpipe/segment/regular` or `/obj/disposalpipespawner/regular` subtypes, it should work fine.

## Why's this needed? 
It's handy, i think it's neat. Makes drawing big long lines of pipes with turns a lot easier. Junctions still have to be done manually though, they have funky behaviours.

### CODE REVIEW

**Comment by frawhst** [comment]
can't wait to make 3 (or heaven forbid, 4) way junctions with these

**Comment by frawhst** [comment]
Also gonna draft this while the checks are screaming

**Comment by TheTyrant25** [comment]
yeah so the actual coding bit appears to be done, it's just uhhhh well bug fixing now

**Comment by TheTyrant25** [comment]
okay looks finished now

File: code/WorkInProgress/recycling/disposal.dm [hunk 1]
```diff
@@ -1884,3 +1889,144 @@ proc/pipe_reconnect_disconnected(var/obj/disposalpipe/pipe, var/new_dir, var/mak
...
+		var/disx = _pipe.x - src.x
+		var/disy = _pipe.y - src.y
+		if (disx == 1 && disy == 0)
+			dpdir |= EAST
+		else if (disx == -1 && disy == 0)
+			dpdir |= WEST
+		else if (disx == 0 && disy == 1)
+			dpdir |= NORTH
+		else if (disx == 0 && disy == -1)
+			dpdir |= SOUTH
```
Comment (pali6) [inline code/WorkInProgress/recycling/disposal.dm:1975]: www.byond.com/docs/ref/info.html#/proc/get_dir

File: code/WorkInProgress/recycling/disposal.dm [hunk 2]
```diff
@@ -1884,3 +1889,144 @@ proc/pipe_reconnect_disconnected(var/obj/disposalpipe/pipe, var/new_dir, var/mak
...
+				dpdir |= EAST
+		else if (disx == -1 && disy == 0)
+			if (_pipe.dpdir & EAST)
+				dpdir |= WEST
+		else if (disx == 0 && disy == 1)
+			if (_pipe.dpdir & SOUTH)
+				dpdir |= NORTH
+		else if (disx == 0 && disy == -1)
+			if (_pipe.dpdir & NORTH)
+				dpdir |= SOUTH
```
Comment (pali6) [inline code/WorkInProgress/recycling/disposal.dm:1991]: ditto

File: code/WorkInProgress/recycling/disposal.dm [hunk 3]
```diff
@@ -1884,3 +1889,144 @@ proc/pipe_reconnect_disconnected(var/obj/disposalpipe/pipe, var/new_dir, var/mak
...
+		CRASH("Lone Pipespawner doesn't connect to anything!\nPipe coords: [src.x] x, [src.y] y, [src.z] z.")
+	else if (length(directions) == 1)
+		// lays a trunk pipe
+		var/obj/disposalpipe/trunk/current = new src.trunk_type(src.loc)
+		current.dir = directions[1]
+		current.dpdir = dpdir
+	else if (length(directions) == 2)
+		// lays a normal pipe segment
+		var/obj/disposalpipe/current = new src.pipe_type(src.loc)
+		if (dpdir & (NORTH + SOUTH) || dpdir & (EAST + WEST))
```
Comment (pali6) [inline code/WorkInProgress/recycling/disposal.dm:2013]:
```suggestion
		if ((dpdir & (NORTH | SOUTH)) || (dpdir & (EAST | WEST)))
```

**Comment by pali6** [comment]
Also, imo these should ignore pipespawners and pipes which are of the wrong type / color etc.
This should allow people to use these spawners next to each other when one is placing mail pipes and the other disposal pipes etc.

**Comment by TheTyrant25** [comment]
> Also, imo these should ignore pipespawners and pipes which are of the wrong type / color etc. This should allow people to use these spawners next to each other when one is placing mail pipes and the other disposal pipes etc.

Yeah that is a good point. They should be able to cross over too

**Comment by TheTyrant25** [comment]
redid the whole detection system so that it includes your three suggestions, @pali6

**Comment by TheTyrant25** [comment]
So the regular pipes work now, but the other kinds don't work at all, plus they break the regular pipes if they're nearby. I really don't know how to fix this

**Comment by TheTyrant25** [comment]
going to completely overhaul the system to use `.type` tomorrow

**Comment by TheTyrant25** [comment]
> I really don't know how to fix this
i found out how to fix it, by the way
just ignore the base type and establish a 'regular' subtype 👍 
and pray the users will use them

File: _std/__build.dm [hunk 4]
```diff
@@ -30,13 +30,13 @@ o+`        `-` ``..-:yooos-..----------..`
...
+#define GOTTA_GO_FAST_BUT_ZLEVELS_TOO_SLOW 1  // Only include the map Atlas, no other zlevels. Boots way faster
 
 //////////// CONVENIENCE OPTIONS FOR TESTING ETC
-//#define DEBUG_EVERYONE_GETS_CAPTAIN_ID // all IDs are captain rank, kept separate from below options to avoid disrupting access-related tests
+#define DEBUG_EVERYONE_GETS_CAPTAIN_ID // all IDs are captain rank, kept separate from below options to avoid disrupting access-related tests
 
-//#define STOP_DISTRACTING_ME //All of the below
+#define STOP_DISTRACTING_ME //All of the below
 
 //#define I_AM_ABOVE_THE_LAW // Prevents all secbots and guardbuddies from spawning, useful for gun testing
```
Comment (pali6) [inline _std/__build.dm:41]: don't commit __build.dm

**Review by pali6 — CHANGES_REQUESTED** [review]
Looks good at a glance except for the __build.dm

File: _std/__build.dm (hunk 4 above)
Comment (TheTyrant25) [reply _std/__build.dm:41]: oops thought i reverted that

File: code/modules/power/cable.dm [hunk 5]
```diff
@@ -573,6 +573,7 @@
 
 	if (length(directions) == 0)
 		cable_laying(0,NORTH)
+		CRASH("Lone cablespawner!\nCable location: [src.x] x, [src.y] y, [src.z] z.")
```
Comment (TheTyrant25) [inline code/modules/power/cable.dm:576]: oh yeah forgot i did this lol. I think it should be fine, single pipe spawners have a similar crash behaviour, so i thought it should match

**Comment by TheTyrant25** [comment]
oops they do not connect to junctions

**Comment by TheTyrant25** [comment]
okay this is all complete now

File: code/modules/power/cable.dm (hunk 5 above)
Comment (TheTyrant25) [reply code/modules/power/cable.dm:576]: just going to remove that actually

File: code/WorkInProgress/recycling/disposal.dm [hunk 6]
```diff
@@ -1889,3 +1921,153 @@ proc/pipe_reconnect_disconnected(var/obj/disposalpipe/pipe, var/new_dir, var/mak
...
+				initialize()
+
+/obj/disposalpipespawner/initialize()
+	var/list/selftile = list()
+	for (var/obj/disposalpipespawner/dupe in range(0, src))
+		if (istype(dupe, src))
+			selftile += dupe
+	if (length(selftile) > 1)
+		CRASH("Multiple pipespawners on coordinate [src.x] x [src.y] y!")
+	qdel(selftile)
```
Comment (frawhst) [inline code/WorkInProgress/recycling/disposal.dm:2005]:
```suggestion
	selftile.Cut()
```
bit cleaner (basically sets the list to 0-length)

**Review by frawhst — APPROVED** [review]
one small thing, otherwise lgtm

File: code/WorkInProgress/recycling/disposal.dm (hunk 6 above)
Comment (frawhst) [reply code/WorkInProgress/recycling/disposal.dm:2005]: actually im just merging this

**Pull request merged**